core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -1079,109 +1103,97 @@ private[spark] class BlockManager(
-
-    // Get a random peer. Note that this selection of a peer is deterministic on the block id.
-    // So assuming the list of peers does not change and no replication failures,
-    // if there are multiple attempts in the same node to replicate the same block,
-    // the same set of peers will be selected.
-    def getRandomPeer(): Option[BlockManagerId] = {
-      // If replication had failed, then force update the cached list of peers and remove the peers
-      // that have been already used
-      if (replicationFailed) {
-        peersForReplication.clear()
-        peersForReplication ++= getPeers(forceFetch = true)
-        peersForReplication --= peersReplicatedTo
-        peersForReplication --= peersFailedToReplicateTo
-      }
-      if (!peersForReplication.isEmpty) {
-        Some(peersForReplication(random.nextInt(peersForReplication.size)))
-      } else {
-        None
-      }
-    }
 
-    // One by one choose a random peer and try uploading the block to it
-    // If replication fails (e.g., target peer is down), force the list of cached peers
-    // to be re-fetched from driver and then pick another random peer for replication. Also
-    // temporarily black list the peer for which replication failed.
-    //
-    // This selection of a peer and replication is continued in a loop until one of the
-    // following 3 conditions is fulfilled:
-    // (i) specified number of peers have been replicated to
-    // (ii) too many failures in replicating to peers
-    // (iii) no peer left to replicate to
-    //
-    while (!done) {
-      getRandomPeer() match {
-        case Some(peer) =>
-          try {
-            val onePeerStartTime = System.currentTimeMillis
-            logTrace(s"Trying to replicate $blockId of ${data.size} bytes to $peer")
-            blockTransferService.uploadBlockSync(
-              peer.host,
-              peer.port,
-              peer.executorId,
-              blockId,
-              new NettyManagedBuffer(data.toNetty),
-              tLevel,
-              classTag)
-            logTrace(s"Replicated $blockId of ${data.size} bytes to $peer in %s ms"
-              .format(System.currentTimeMillis - onePeerStartTime))
-            peersReplicatedTo += peer
-            peersForReplication -= peer
-            replicationFailed = false
-            if (peersReplicatedTo.size == numPeersToReplicateTo) {
-              done = true  // specified number of peers have been replicated to
+    val numPeersToReplicateTo = level.replication - 1
+
+    @tailrec def replicateBlock(
+      numFailures: Int,
+      peersForReplication: Seq[BlockManagerId],
+      peersReplicatedTo: Set[BlockManagerId],
+      peersFailedToReplicateTo: Set[BlockManagerId]): Set[BlockManagerId] = {
+
+      if (numFailures > maxReplicationFailures
+                || peersForReplication.isEmpty
+                || peersReplicatedTo.size == numPeersToReplicateTo) {
```

Comment:
Although I cannot find a documentation for indentation about this, it might be better be consistent because I see you added

```scala
executorId == id.executorId &&
  port == id.port &&
  host == id.host &&
  topologyInfo == id.topologyInfo
```

So this might have to be

```scala
if (numFailures > maxReplicationFailures
    || peersForReplication.isEmpty
    || peersReplicatedTo.size == numPeersToReplicateTo) {
```